fix: set index and value to 0 for array_get with predicate

guipublic opened the pull request:
# Description

## Problem\*

Resolves issue on noir-protocol-circuit and should also resolve issues described in https://github.com/noir-lang/noir/pull/4716.
The issue it resolves is that array_get under a predicate are done at index 0 if the predicate is false, but since arrays are not homogenous, it may contain some value that overflow the type of the array_get.

## Summary\*
When getting value from an array_get, I multiply it with the predicate to avoid any possible overflow if element at index 0 has not the same type as the expected one.

## Additional Context
If the array is simple (not nested), I get the offset to a compatible type and use it when computing the predicate_index.
If the first element of the array has compatible size, then the predicate_index will be correct
If not, we fallback to the multiplication of the value with the predicate.


## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

vezenovm commented on compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:
```diff
@@ -1200,7 +1200,14 @@ impl<'a> Context<'a> {
         );
-        let value = self.array_get_value(&res_typ, block_id, &mut var_index)?;
+        let mut value = self.array_get_value(&res_typ, block_id, &mut var_index)?;
+        if let AcirValue::Var(value_var, typ) = value {
```
What would happen if this is `AcirValue::Array`?

guipublic replied:
My thinking was that overflows would only matter for numerical values, which is the case that is handled.

vezenovm replied:
Ok I am going to approve for now as to not block it. If I succeed in making a failure for the array case it can be a separate PR

TomAFrench commented:
It would be interesting to compare this against the solution of adding an index offset instead of reading 0 to ensure that we read the correct typed value from the array. This keeps the final value a single witness rather than a quadratic expression so should result in better codegen in later opcodes.

TomAFrench commented:
Edit: the above comment isn't a blocker for this PR and I've opened #4972 to follow up on this.

TomAFrench commented:
It also seems like we can be smarter about when to multiply in the predicate/add an offset to the index. We only need to do this when the type which we would read is larger than the type we want to read. We can then avoid this in two cases.

1. The array is homogeneous (I think this would cut out 99% of the circuit size increase we see)
2. The array is heterogeneous but the type of the first field of the element type is smaller than the type we want to read. i.e. `[Struct {foo: bool, bar: Field}; 3]` can always read the first field in the flattened array safely and use its value for all other indices.

guipublic commented:
> It also seems like we can be smarter about when to multiply in the predicate/add an offset to the index. We only need to do this when the type which we would read is larger than the type we want to read. 

Good point, I have also added the use of the offset when the array is not a nested array. So the rule now is:
- if array is not nested, find the offset and use it for the predicate_index, else
- if the first element of the array has compatible bit_size, we can use directly the predicate_index, else
- fallback to the multiply the value with the predicate.

I have updated the PR description.
I could not get the offset to work with nested arrays but it is definitively do-able. However, almost all cases are now properly optimised.

vezenovm commented on compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:
```diff
@@ -956,14 +956,42 @@ impl<'a> Context<'a> {
+            if let Type::Array(item_type, _) = array_typ {
+                // For simplicity we compute the offset only for simple arrays
+                let mut contains_an_array = false;
+                for typ in item_type.iter() {
```
We have `can_omit_element_sizes_array` that checks whether we have a nested array or slice. We can check against that on its own as to remove this loop

guipublic replied:
Done

vezenovm commented on compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:
```diff
@@ -956,14 +956,42 @@ impl<'a> Context<'a> {
+        let array_typ = dfg.type_of_value(array_id);
+        if dfg.instruction_results(instruction).len() == 1 {
+            let result_type = dfg.type_of_value(dfg.instruction_results(instruction)[0]);
+            if let Type::Array(item_type, _) = array_typ {
```
```suggestion
            if let Type::Array(item_type, _) | Type::Slice(types) = array_typ {
```
We probably want to look at slices too. We banned nested slices but a slice can have nested arrays

guipublic replied:
Done

vezenovm commented on compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:
```diff
@@ -1204,7 +1242,28 @@ impl<'a> Context<'a> {
+        if let AcirValue::Var(value_var, typ) = &value {
+            let array_id = dfg.resolve(array_id);
+            let array_typ = dfg.type_of_value(array_id);
+            if let Type::Numeric(numeric_type) = array_typ.first() {
+                if let AcirType::NumericType(num) = typ {
```
```suggestion
            if let (Type::Numeric(numeric_type), AcirType::NumericType(num)) = (array_typ.first(), typ) {
```
We should be able to combine these if statements with something like this

guipublic replied:
Done

vezenovm commented on compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:
```diff
@@ -1063,14 +1092,21 @@ impl<'a> Context<'a> {
-            self.acir_context.mul_var(index_var, self.current_side_effects_enabled_var)?;
+        let predicate_index = if offset != 0 {
+            let offset = self.acir_context.add_constant(offset);
+            let sub = self.acir_context.sub_var(index_var, offset)?;
```
Could we add some comments as to why we are subtracting the offset to multiply by the predicate and then adding the offset back?

guipublic replied:
there is already a comment:
`if an offset is specified, predicate_index is index*predicate + (1-predicate)*offset`
do you want more explanation?

vezenovm replied:
Ah I see now that is fine then

vezenovm commented on compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:
```diff
@@ -956,14 +956,42 @@ impl<'a> Context<'a> {
+            let result_type = dfg.type_of_value(dfg.instruction_results(instruction)[0]);
+            if let Type::Array(item_type, _) = array_typ {
+                // For simplicity we compute the offset only for simple arrays
+                let mut contains_an_array = false;
+                for typ in item_type.iter() {
+                    if typ.contains_an_array() {
+                        contains_an_array = true;
+                        break;
+                    }
+                }
+                if !contains_an_array {
+                    for (i, typ) in item_type.iter().enumerate() {
+                        if result_type == *typ {
+                            offset = Some(i);
+                            break;
+                        }
+                    }
+                }
```
We can use `can_omit_element_sizes_array` as well here which does this `any` check on an array or slice's internal types

guipublic replied:
I went for can_omit_element_sizes_array()

guipublic commented on compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:
```diff
@@ -1171,12 +1207,14 @@ impl<'a> Context<'a> {
         array: ValueId,
         mut var_index: AcirVar,
         dfg: &DataFlowGraph,
+        mut side_effect_free: bool,
```
Done

TomAFrench reviewed (approved):
Some nits but good to go.

guipublic commented on compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:
```diff
@@ -957,13 +957,39 @@ impl<'a> Context<'a> {
+                    }
+                }
+            }
+        }
```
Done

guipublic commented on compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:
```diff
@@ -1171,12 +1205,14 @@ impl<'a> Context<'a> {
     }
 
     /// Generates a read opcode for the array
+    /// index_side_effect false means that we ensured index will have a type matching the value in the array
```
Done

guipublic commented on compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:
```diff
@@ -1063,14 +1090,21 @@ impl<'a> Context<'a> {
+            self.acir_context.add_var(pred, offset)?
+        } else {
+            self.acir_context.mul_var(index_var, self.current_side_effects_enabled_var)?
+        };
```
Done

guipublic commented on compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:
```diff
@@ -1054,6 +1080,7 @@ impl<'a> Context<'a> {
     ///     in case we have a nested array. The index for SSA array operations only represents the flattened index of the current array.
     ///     Thus internal array element type sizes need to be computed to accurately transform the index.
     /// - predicate_index is 0, or the index if the predicate is true
+    /// if an offset is specified, predicate_index is index*predicate + (1-predicate)*offset
```
Done
